intro-to-functions/defining-functions.md

```diff
@@ -0,0 +1,458 @@
+    print("Wow!")
+    return a+b
+```
+
+Function bodies can read their parameters and use them like variables. They can use these variables and determine a return value.
+```python
+def add(a, b):
+    doubled_a = a*2
+    smaller_b = b-1
+    sum = a+b
+    print(f"Just for fun, I can print and use my variables {doubled_a} and {smaller_b}")
+    return sum
+```
+
+Functions can have nested logic, and conditional return values.
+```python
+def add(a, b):
+    if a < 0:
+        print("I only want to add if a is greater than zero! I'll return false instead.")
+        return false
+    elif b == a:
+        print("Did you know that a and b are the same value?")
+    elif a % 2 == 0:
+        print("Did you know that a is an even number?")
+        if b % 2 == 0:
+            print("Did you know that b is also an even number?!?!?!?")
+            print("Jackpot!")
+            return 777
+
+    return a+b
+```
+
+Function bodies can choose to not return anything, and will instead return `None`.
+```python
+def add(a, b):
+    print("I don't need to return anything.")
+    print("If I don't, I'll return None automatically.")
+    print("Instead, I'll just print some addition.")
+    print(a+b)
+    print("Great!")
+    print("I did my job.")
+    print("I'll get going now.")
+```
+
+### Functions With an Empty Body with `pass`
+
+The `pass` statement is Python statement that executes no code. For example:
+
+```python
+def calculate_bill(items, sales_tax, tip_percentage):
+    pass
+```
+
+[This is useful when we developers want to write function signatures that don't do anything at that moment](https://en.wikipedia.org/wiki/Method_stub), but Python syntax requires function bodies to be at least one line.
+
+## Calling Your Functions In Your Python Code
+
+Now that we can define functions, we can apply our knowledge about invoking functions to them!
+
+Remember, invoking a function means:
+- Knowing the function's name
+- Knowing the function's parameters, and what arguments we want to pass into them
+- Knowing the syntax to invoke the function: `()`!
+
+```python
+def sing_happy_birthday(name):
+    print("Happy birthday to you...")
+    print("Happy birthday to you...!")
+    print(f"Happy birthday, dear {name}")
+    print("Happy birthday to you!")
+
+sing_happy_birthday("Melinda")
+sing_happy_birthday("Kerry")
+```
+
+### Order Still Matters
+
+Order still matters. We cannot invoke a function before our Python interpreter knows that it's defined.
+
+Predict: What error we will see and why we will see it?
+
+```python
+sing_happy_birthday("Melinda")
+sing_happy_birthday("Kerry")
+
+def sing_happy_birthday(name):
+    print("Happy birthday to you")
+    print("Happy birthday to you")
+    print(f"Happy birthday, dear {name}")
+    print("Happy birthday to you")
+```
+
+Answer: We will get `NameError: name 'sing_happy_birthday' is not defined` because we try to invoke the `sing_happy_birthday` function before it's defined.
```

Comment:
Wait a minute was this meant to be a Learn question, or as a `read-a-long` (sorta) rhetorical question? Anyway, here's a Learn MC option just in case!
```suggestion
### !challenge

* type: multiple-choice
* id: e205de80-0a88-486d-a842-30a29d5f896e
* title: Function Error

##### !question

Predict: What error we will see and why we will see it?

```python
sing_happy_birthday("Melinda")
sing_happy_birthday("Kerry")

def sing_happy_birthday(name):
    print("Happy birthday to you")
    print("Happy birthday to you")
    print(f"Happy birthday, dear {name}")
    print("Happy birthday to you")
```

##### !end-question

##### !options

* ValueError
* TypeError
* NameError

##### !end-options

##### !answer

* NameError
##### !end-answer


#### !explanation 
We will get `NameError: name 'sing_happy_birthday' is not defined` because we try to invoke the `sing_happy_birthday` function before it's defined.
#### !end-explanation

### !end-challenge
```